src/modules/pages/NewOAuthServerPage/partials/OAuthClientEndpoints/OAuthClientEndpoint.js

```diff
@@ -15,204 +16,272 @@ import Hint from '../../../../labels/Hint/Hint';
     name: PropTypes.string.isRequired,
+    schema: PropTypes.object.isRequired,
 };
 
-const OAuthClientEndpoint = ({ endpoint, name, schema }) => {
-    const createOptions = (list1, list2) => {
-        const combinedListOfUnitsAndLabels = R.zip(list1, list2);
+class OAuthClientEndpoint extends PureComponent {
+    state = {
+        upstreams: this.props.initialValues[this.props.category][this.props.name].upstreams || {}, // fallback for old endpoints (they have `upstreams: null`), probably temporary
+    };
 
-        return combinedListOfUnitsAndLabels.map(item => ({
+    createStrategyOptions = list => {
+        const extractNames = list => list.map(item => item.balancing);
+        const labelCombiner = item => ({
             label: item[1],
-            value: item[0],
-        }));
+            value: item[1],
+            options: item[0],
+        });
+        const combineListOfUnitsAndLabels = list => list.map(labelCombiner);
+
+        return R.compose(
+            combineListOfUnitsAndLabels,
+            R.zip(list),
+            extractNames,
+        )(list);
     };
-    const optionsTransformer = config => config.map(item => ({
-        label: item,
-        value: item,
-    }));
 
-    return (
-        <div className={b('section')}>
-            <div className={b('section-title')}>{name}</div>
-            <div className={row({fullwidth: true}).mix('j-api-form__row')}>
-                <div className={row('item')}>
-                    <div className={col()}>
-                        <div className={col('item')}>
-                            <Label>Listen Path</Label>
+    handleChangeStrategy = value => {
+        this.setState(() => ({
+            upstreams: {
+                balancing: value.value,
+                targets: value.options,
+            }
+        }), () => {
+            this.props.change(`${this.props.category}.${this.props.name}.upstreams.balancing`, value.value);
+        });
+    };
+
+    renderStrategy = balancing => {
+        switch (balancing) {
+            case 'roundrobin': {
+                return (
+                    <RoundrobinTargets
+                        name={`${this.props.category}.${this.props.name}.upstreams.targets`}
+                        title="Roundrobin targets"
+                    />
+                );
+            }
+            case 'weight': {
+                return (
+                    <WeightTargets
+                        name={`${this.props.category}.${this.props.name}.upstreams.targets`}
+                        title="Weight targets"
+                    />
+                );
+            }
+            default:
+                return null;
+        }
+    };
+
+    render() {
+        const { category, endpoint, name, schema } = this.props;
+        const createOptions = (list1, list2) => {
+            const combinedListOfUnitsAndLabels = R.zip(list1, list2);
+
+            return combinedListOfUnitsAndLabels.map(item => ({
```

Comment:
```js
const asLabelValue = item => ({ label: item[1], value: item[0] })
return R.zip(list1, list2).map(asLabelValue)
```

Reply:
Again my bad... this code is redundant here... 🔫